session/tidb.go

```diff
@@ -171,11 +171,20 @@ func runStmt(ctx context.Context, sctx sessionctx.Context, s sqlexec.Statement)
 		// If the user insert, insert, insert ... but never commit, TiDB would OOM.
 		// So we limit the statement count in a transaction here.
 		history := GetHistory(sctx)
-		if history.Count() > int(config.GetGlobalConfig().Performance.StmtCountLimit) {
-			err1 := se.RollbackTxn(ctx)
-			terror.Log(errors.Trace(err1))
-			return rs, errors.Errorf("statement count %d exceeds the transaction limitation, autocommit = %t",
-				history.Count(), sctx.GetSessionVars().IsAutocommit())
+		perfConfig := config.GetGlobalConfig().Performance
+		if history.Count() > int(perfConfig.StmtCountLimit) {
+			if !perfConfig.BatchCommit {
+				err1 := se.RollbackTxn(ctx)
+				terror.Log(errors.Trace(err1))
+				return rs, errors.Errorf("statement count %d exceeds the transaction limitation, autocommit = %t",
+					history.Count(), sctx.GetSessionVars().IsAutocommit())
+			}
+			err = se.NewTxn()
+			// The transaction does not committed yet, we need to keep it in transaction.
+			// The last history could not be "commit" statement.
+			// It means it is impossible to start a new transaction at the end of the transaction (the "commit" statement).
+			// Because after the server executed "commit" statement, the session is out of the transaction.
+			se.sessionVars.SetStatusFlag(mysql.ServerStatusInTrans, true)
```

Comment:
What will happen if the last statement is a `rollback`, will there be a corner case that this line put the session into a new transaction while it should be rollbacked ?

Reply:
`rollback` is the same as `commit`.

Reply:
I will update the comment and add a `rollback` test case.

Reply:
```
s1:
set @@autocommit = 0;
insert into t value (1);
rollback;

s2:
insert into t values (1);
insert into t values (2);

s1:
select  * from t;    // what it see ?
```

@jackysp

Reply:
So the statement count limit is 2 for this case?